### DISCUSSION TITLE
如何使用Alt快捷键快捷输入字符

### DISCUSSION THREAD

**Discussion by Jiang10086** [rime/home#1552] [Q&A]
比如，键盘上很难打出En dash（–）和Em dash（—），所以我想使用<kbd>Alt</kbd> + <kbd>-</kbd>直接打出En dash（–），用<kbd>Alt</kbd>+<kbd>Shift</kbd>+<kbd>-</kbd>打出Em dash（—），不知道单用小狼毫不用其他工具该怎么设置呢？

**Comment by lotem** [comment]
非常實用的功能。可惜目前還沒有這樣的實現，只能迂迴，用按鍵綁定的方式把組合鍵轉化爲字符按鍵序列。

**Reply by Jiang10086** [reply]
请问具体实现思路是怎样的呢？我用RIME（小狼毫）不久，不太清楚各项配置。

**Reply by lotem** [reply]
配置比較複雜呀。先講講思路吧。

首先，標點符號，`punctuator` 可以設置爲多選。比如默認設置裏 <kbd>/</kbd> <kbd>[</kbd> <kbd>]</kbd> 等按鍵會對應一組符號，展示在候選列表裏，可知：按鍵序列 `/1` 上屏 `、`，按鍵序列 `/2` 上屏 `/` 按鍵序列 `/4` 上屏 `÷` 等等。 <kbd>-</kbd> 只設置爲對應 `-`，首先要把它改成多選的。考慮到 `-` 比較常用，也可以把 en dash, em dash 放在 <kbd>_</kbd> 的候選裏，和破折號 `——` 並列。

設置了以上符號，下一步是 `key_binder` 的配置，把相應快捷鍵映射爲按鍵序列（`send_sequence`）如 `_2`、`_3` 等等。如所求。
唯一副作用是破折號不會直接上屏，需要加空格上屏。

**Reply by Jiang10086** [reply]
这样的话，en dash、em dash能直接上屏吗？

**Reply by lotem** [reply]
能，發送選字的數字鍵就上屏了。

**Reply by Jiang10086** [reply]
感谢你提供的思路，但我想要的是直接上屏（就比如shift+-打出破折号），而不需要选字。

**Comment by popyoung** [comment]
我也有个类似的需求，是想在中文输入情况下用 alt 或 ctrl + 逗号输出英文逗号。试了下`- { when: always, accept: Control+comma, send_sequence: "{Shift_L},{Shift_L}" }` ，好像和我理解的不一样，不知为何 shift 并没有被执行。然后又试了`- { when: always, accept: Control+comma, send: Shift }`也没有成功切换中英文，是不是这里必须使用 toggle ？lua 能实现我说的功能吗？

**Comment by popyoung** [comment]
最后我自己用Lua实现了，详见 https://github.com/iDvel/rime-ice/issues/1134

**Comment by lotem** [comment]
別用 Lua。

用 `punctuator` 定義符號的輸入碼，如 `/bd`
再用 `key_binder` 將快捷鍵綁定爲發送（`send_sequence`）符號的輸入碼（及選字碼）。
按我的配置打出「―」要翻頁一次，即發送 `/bd.8`

**Reply by popyoung** [reply]
那个帖子里也有位朋友建议我别用 lua，他建议我使用 ahk，为什么啊？
当然你提供的思路也很好，学习了。

**Reply by popyoung** [reply]
> 別用 Lua。
> 
> 用 `punctuator` 定義符號的輸入碼，如 `/bd` 再用 `key_binder` 將快捷鍵綁定爲發送（`send_sequence`）符號的輸入碼（及選字碼）。 按我的配置打出「―」要翻頁一次，即發送 `/bd.8`

我试了一下send_sequence，和 lua 实现方案类似的是 Ctrl+) 无法捕获，Lua 在 Processor 最前方输出日志看会发现 crtl+shift+) 只有release，没有 press。这个还是要用 lua 写一个对应的特殊处理才能勉强解决。不知道是小狼毫的 bug 还是什么。
```
  "key_binder/bindings/+":
    - { when: always, accept: Control+comma, send_sequence: "vesym1" }
    - { when: always, accept: Control+period, send_sequence: "vesym2" }
    - { when: always, accept: Control+Shift+colon, send_sequence: "vesym3" }
    - { when: always, accept: Control+semicolon, send_sequence: "vesym4" }
    - { when: always, accept: Control+Shift+quotedbl, send_sequence: "vesym5" }
    - { when: always, accept: Control+apostrophe, send_sequence: "vesym{Next}1" }
    - { when: always, accept: Control+bracketleft, send_sequence: "vesym{Next}2" }
    - { when: always, accept: Control+bracketright, send_sequence: "vesym{Next}3" }
    - { when: always, accept: Control+Shift+braceleft, send_sequence: "vesym{Next}4" }
    - { when: always, accept: Control+Shift+braceright, send_sequence: "vesym{Next}5" }
    - { when: always, accept: Control+Shift+parenleft, send_sequence: "vesym{Next}{Next}1" }
    - { when: always, accept: Control+Shift+parenright, send_sequence: "vesym{Next}{Next}2" }
    - { when: always, accept: Control+Shift+less, send_sequence: "vesym{Next}{Next}3" }
    - { when: always, accept: Control+Shift+greater, send_sequence: "vesym{Next}{Next}4" }
    - { when: always, accept: Control+Shift+question, send_sequence: "vesym{Next}{Next}5" }
    - { when: always, accept: Control+Shift+exclam, send_sequence: "vesym{Next}{Next}{Next}1" }
    - { when: always, accept: Control+backslash, send_sequence: "vesym{Next}{Next}{Next}2" }
    - { when: always, accept: Control+Shift+dollar, send_sequence: "vesym{Next}{Next}{Next}3" }
    - { when: always, accept: Control+Shift+asciicircum, send_sequence: "vesym{Next}{Next}{Next}4" }```
```
